### PR TITLE
perf(txn): de-duplicate the context keys and predicates

### PR DESCRIPTION
**Pull request by NamanJain8**
The zero raft loop got stuck because the proposals took a long time to process. Hence, the goroutine which checks the quorum gets struck leading to leadership change (even in single zero single alpha). 

**Update:** The real issue was the Keys in txnContext. While merging the conflict keys, deduplication was not done. Do deduplication before sending the request to zero. 

The issue is due to lock contention.
https://github.com/dgraph-io/dgraph/blob/c66e86985b642eb6b563014ed05389e7b3170f72/dgraph/cmd/zero/oracle.go#L132-L151 is holding the lock, while this function tries to send update https://github.com/dgraph-io/dgraph/blob/c66e86985b642eb6b563014ed05389e7b3170f72/dgraph/cmd/zero/oracle.go#L273-L283.

The processing of the keyCommit map is slow and the locks for o.commits and o.keyCommit can be separated by using `sync.Map` for o.commits.

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7478)

### CODE REVIEW

File: dgraph/cmd/zero/oracle.go [hunk 1]
```diff
@@ -130,8 +143,8 @@ func (o *Oracle) purgeBelow(minTs uint64) {
 }
 
 func (o *Oracle) commit(src *api.TxnContext) error {
-	o.Lock()
-	defer o.Unlock()
+	o.keyCommitLock.Lock()
```
Comment (jarifibrahim) [inline dgraph/cmd/zero/oracle.go:146]: This might not work.

**Review by manishrjain — COMMENTED** [review]
Move the proposal application out of the Zero Raft loop, into an applyCh style application, like we do in Alpha. That would tackle the slowness issues.

> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7478#-:-MUiNRJa0R8l_5xFBd3Y:b-mz586f)* status: 0 of 1 files reviewed, 1 unresolved discussion (waiting on @NamanJain8 and @vvbalaji-dgraph)

**Review by NamanJain8 — COMMENTED** [review]
As discussed the real issue was with txnContext.Keys containing duplicates. So, we don't need that Raft loop change.

> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7478#-:-MUoApHJ0veXH5yHBOab:b-b6iaau)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @martinmr, @NamanJain8, and @vvbalaji-dgraph)

**Review by jarifibrahim — APPROVED** [review]
Nice work @NamanJain8 :tada:

**Pull request merged**